rp2040-hal/src/spi.rs

```diff
@@ -88,15 +92,17 @@ impl Sealed for u8 {}
 impl Sealed for u16 {}
 
 /// Spi
-pub struct Spi<S: State, D: SpiDevice, const DS: u8> {
+pub struct Spi<S: State, D: SpiDevice, P: ValidSpiPinout<D>, const DS: u8> {
```

Comment:
With the new parameter `P`, how can I translate such a type alias to the new API?
```
type SpiBus = hal::Spi<hal::spi::Enabled, pac::SPI0, 8>
```
(Taken from https://github.com/Neotron-Compute/Neotron-Pico-BIOS/blob/develop/src/main.rs#L112)

Reply:
It depends on whether you man the pinId to be compile time checked.
If not, you'd probably do something like:

```rust
type SpiPins = (
    Pin<DynPinId, FunctionSpi, DynPullType>, // MOSI
    Pin<DynPinId, FunctionSpi, DynPullType>, // MISO
    Pin<DynPinId, FunctionSpi, DynPullType>, // SCLK
);

type SpiBus = hal::Spi<hal::spi::Enabled, pac::SPI0, SpiPins, 8>;
```

Reply:
Which makes me realize that I should've added a `try_new` function to runtime check the valid Pin Function on serial, spi and i2c like I did for Adc 🤔 .

I'll impl that asap (probably in the next couple of days).

Reply:
I updated the PR and it will now be:

```rust
type SpiPins = (
    ValidatedPinTx<Pin<DynPinId, FunctionSpi, DynPullType>, SPI0>, // MOSI
    ValidatedPinTx<Pin<DynPinId, FunctionSpi, DynPullType>, SPI0>, // MISO
    ValidatedPinTx<Pin<DynPinId, FunctionSpi, DynPullType>, SPI0>, // SCLK
);

type SpiBus = hal::Spi<hal::spi::Enabled, pac::SPI0, SpiPins, 8>;
```

I'm not sure it is the most elegant but it does the trick.